### PR TITLE
Add aesara-base package and clean up dependencies

### LINKED ISSUE #76: Add aesara-base package
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

In `pymc` we have a `pymc-base` package that does not have requirements on compilers or `blas`. This would be very useful for `aesara` too incase I don't want to chose my own computational backend.

### Installed packages

```shell
foo
```

### Environment info

```shell
foo
```

### PR DESCRIPTION
**Pull request by maresb**
Closes #76

Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [X] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
CI failure was just due to a connection issue. Retrying...

**Comment by maresb** [comment]
This is unfortunately being blocked by https://github.com/conda-forge/mkl-service-feedstock/issues/24, because this recipe builds for PyPy 3.7, and it's impossible to install mkl-service and PyPy 3.7 simultaneously.

**Comment by twiecki** [comment]
Can we drop pypy 3.7?

**Comment by maresb** [comment]
I also had that thought. There is a newer migration for pypy 3.8 & 3.9, so there doesn't seem to be much point in continuing to offer 3.7. I will ask on Gitter if there's a procedure for this, so that we don't confuse their migration infrastructure...

https://gitter.im/conda-forge/conda-forge.github.io?at=62c06471b408e830dc7a06c2

**Comment by maresb** [comment]
I found a selector for pypy which I was previously unaware of. Now I can set pypy3.7 to openblas. :smile:

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
Cool, all green! :tada:

**Comment by maresb** [comment]
We should probably discuss the extra stuff to add to `aesara-base` for `aesara`. Currently I put:

```
        - {{ compiler('c') }}  # [not win]
        - cxx-compiler  # [not win]
        - {{ compiler('m2w64_c') }}  # [win]
        - {{ compiler('m2w64_cxx') }}  # [win]
        # Use openblas on Apple M1 or pypy3.7. For everything else, use mkl-service.
        - openblas  # [arm64 or (python_impl == 'pypy' and py==37)]
        - mkl-service  # [not (arm64 or (python_impl == 'pypy' and py==37))]
```

Is this desired, or do we want something else?

**Comment by twiecki** [comment]
> Is this desired, or do we want something else?

That looks good to me!

File: recipe/meta.yaml
Comment (twiecki) [inline recipe/meta.yaml:42]: I'm not sure even this (and the next two) are a hard dependency. Let's try moving them.

Comment (maresb) [reply recipe/meta.yaml]: Good point.

Host dependencies confuse me. My understanding is that they are essentially build dependencies, however in the case of cross-compiling, they reference the target system rather than the build system. They should not be runtime dependencies.

Looking in the commit history, `libblas` and `liblapack` were added by @brandonwillard in #27. I wonder what was the rationale behind that. If Aesara compiles some libraries at build time, and these libraries require BLAS to be full-featured, then we'll definitely want to add them back. However, if the goal was to add them as runtime dependencies, then I think this PR accomplishes that.

Comment (twiecki) [reply recipe/meta.yaml]: I don't think libblas is needed for building aesara, but it's certainly needed for running (if the user wants to C-compile).

Comment (twiecki) [reply recipe/meta.yaml]: But we do need them for `aesara` still.

**Comment by maresb** [comment]
Todo: the ci infrastructure is not capable of running tests on the cross-compiled osx-arm64 packages, so these should ideally be tested by hand when we are ready. The builds can be downloaded under checks -> Azure pipelines as "artifacts".

File: recipe/meta.yaml
Comment (maresb) [reply recipe/meta.yaml]: I checked each individual CI run, and in each case libblas and liblapack are being installed as implicit dependencies, presumably by mkl-service or openblas.

We could still add them as explicit runtime dependencies if that makes sense.

I wonder what the `blas` package does...

Comment (maresb) [reply recipe/meta.yaml]: Once we understand exactly what the requirements are, we may want to consult the conda-forge experts with blas experience to check this.

Here is the [advice from the knowledge base](https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=mesa#blas), but i don't think this is directly applicable to us since we require _runtime_ building against blas.

Comment (maresb) [reply recipe/meta.yaml]: This situation reminds me of [cmdstanpy](https://github.com/conda-forge/cmdstan-feedstock/blob/main/recipe/meta.yaml) but there we don't need blas.

Comment (twiecki) [reply recipe/meta.yaml]: Yeah, `blas` is required for some odd reason, we've seen compile errors without it (https://discourse.pymc.io/t/pymc-4-0-0-b2-fatal-error-cblas-h/8791).

Comment (maresb) [reply recipe/meta.yaml]: Thanks, that's really helpful! In order to make this info easy to reference, I opened #80. I'm thinking we can explain our reasoning as comments in the feedstock and add such links as references. That way we don't rely so much on folklore.

File: recipe/meta.yaml
Comment (twiecki) [inline recipe/meta.yaml:49]: Is libpython still required here? Maybe also move this down to `aesara`?

Comment (maresb) [reply recipe/meta.yaml:49]: Good question...

Oh... I just checked, and the dependencies in general seem to have not been updated in a long time! :joy:

As I mentioned a few weeks ago (I think regarding the PyMC feedstock), the Conda-Forge infrastructure doesn't automatically update the dependencies (yet), because this is an extremely difficult problem.

I'll have a look...

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
Weird!!! The Python executable suddenly can't be found for `osx_arm64`. Maybe something went wrong after I merged main and rerendered?

**Comment by maresb** [comment]
According to the [diff](https://github.com/conda-forge/aesara-feedstock/compare/51b8498..855c92f) before and after merging main and rerendering, it looks like I lost something in the CI configuration regarding libblas and liblapack. Probably I should have been rerendering more often. Could this really be the source of the missing python executable though?

**Comment by maresb** [comment]
Open question: why does `cxx-compiler` work but `compiler('cxx')` does not?

**Comment by maresb** [comment]
I opened two parallel draft PRs #81 and #82 to try to get to the bottom of the missing python executable on osx_arm64. I did this because each PR can run the CI independently.

File: recipe/meta.yaml
Comment (maresb) [reply recipe/meta.yaml]: Dependencies updated, and libpython is now gone, so marking as resolved.

**Comment by maresb** [comment]
It turns out that the issue with the failing builds was indeed a bug with the Conda-Forge infrastructure. :sweat_smile:
See [on Gitter](https://gitter.im/conda-forge/conda-forge.github.io?at=62c1b0e71227f62be3107ade).

**Comment by twiecki** [comment]
Thanks for your work on this, the meta.yml file is shaping up nicely. So `libpython` is not required at all then?

**Comment by maresb** [comment]
@twiecki thanks! And everything is green again! :tada: 

It seems that `libpython` is some sort of Windows-specific package. It's unnecessary to list it as a direct dependency because it occurs in all Windows builds as an indirect dependency. (Unfortunately I'm having trouble determining from what.) It was added in https://github.com/conda-forge/aesara-feedstock/pull/29.

So if we have a particular reason that we need `libpython` we should probably list it as a direct dependency, but otherwise it seems fine to me as-is.

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
@twiecki, I'm fairly satisfied with how this looks. Let's get comments from @brandonwillard. If he is satisfied, then let's seek a review from a BLAS expert on Conda-Forge. Perhaps we could ping `conda-forge/blas`?

**Comment by twiecki** [comment]
Awesome, I agree that this looks much better. Let's wait on @brandonwillard.

**Comment by twiecki** [comment]
> It seems that libpython is some sort of Windows-specific package. It's unnecessary to list it as a direct dependency because it occurs in all Windows builds as an indirect dependency.

Could you explain a bit more why it's not needed explicitly?

I just chatted with @brandonwillard and he's fine with the approach. Just need to make sure we didn't break anything with installing the compiler chain and necessary libraries, but I guess the tests ensure that?

**Comment by maresb** [comment]
> Could you explain a bit more why it's not needed explicitly?

Some other package pulls in `libpython` regardless of whether or not we do it explicitly. Thus in the end it makes no difference, unless whatever it was stops requiring it.

Since you're concerned, let's add it in with a comment. There's no harm in it, so if it increases confidence then let's do it.

The tests only check for errors / warnings on import. The tests do not run for osx-arm64.

**Comment by twiecki** [comment]
> Since you're concerned, let's add it in with a comment. There's no harm in it, so if it increases confidence then let's do it.

No, it's fine, just wanted to understand.

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks so much @maresb, this really made the package a lot better!

**Comment by brandonwillard** [comment]
> Some other package pulls in `libpython` regardless of whether or not we do it explicitly. Thus in the end it makes no difference, unless whatever it was stops requiring it.

Since Aesara uses the Python C libraries and headers directly, this is not a sound approach.  

Generally, any time a codebase explicitly relies on a library, it should have that library as a direct dependency, even if that dependency is known to be provided by other dependencies; otherwise, one invites avoidable issues caused by changes in said dependencies (e.g. when a dependency removes its dependence on your indirect dependency, sets its own install options for the indirect dependency, etc.)

**Comment by maresb** [comment]
> Since Aesara uses the Python C libraries and headers directly, this is not a sound approach.

Exactly, and because I was unsure whether or not this was a direct dependency I wanted your to review. :smiley:

But no harm, just a bit of extra work for me. We can fix this going forward in #88, and retroactively in #87.

**Comment by maresb** [comment]
Another important point is that the osx-arm64 packages have, to my knowledge, not been tested. Could someone please test `aesara` and `aesara-base` on M1?

**Comment by twiecki** [comment]
Just sampled a model and that went fine.

**Pull request merged**
